src/element-service.js

```diff
@@ -185,31 +191,47 @@ function assertService(service, id, extension) {
 }
 
 /**
- * Waits for an extension if a stub is present
- * @param {!Window} win
+ * Get and cache a list of all the extension JS files
+ * @param {Document} doc
+ * @return {!Array<string>}
+ * @private
+ */
+function getExtensions(doc) {
+  dev().assert(doc.body,
+      'getExtensions() must only be called after document body is available');
+  if (customElements_ !== null) {
+    return customElements_;
+  }
+  customElements_ = [];
+  const list = doc.head.querySelectorAll('script[custom-element]');
+  for (let i = 0; i < list.length; i++) {
+    customElements_.push(list[i].getAttribute('custom-element'));
+  }
+  return customElements_;
+}
```

Comment:
We can avoid duping this query code by extracting the logic out of `stubElementsForDoc` into a helper function.

Reply:
I thought about that - the use case is different `stubElementsforDoc` needs a non cached query and we are intentionally caching.   The only duped code is the `querySelectorAll`, is it worth it for that?

Reply:
Minor impact but nice IMO. It should reduce the bundle size by a tad. 

```js
export function extensionScriptsInNode(node) {
  const extensions = [];
  const list = node.querySelectorAll('script[custom-element]');
  for (let i = 0; i < list.length; i++) {
    const name = list[i].getAttribute('custom-element');
    extensions.push(name);
  }
  return extensions;
}

export function stubElementsForDoc(ampdoc) {
  const extensions = extensionScriptsInNode(ampdoc.getHeadNode());
  extensions.forEach(name => {
    ampdoc.declareExtension(name);
    stubElementIfNotKnown(ampdoc.win, name);
  });
}
```

On second thought, we probably can't cache this without proper scoping (normal AMP doc vs. PWA vs. FIE). The first two cases would be solved by putting the cache on the `AmpDoc` object.